File: questions/230.KthSmallestElementInABST.go
package questions

func kthSmallest(root *TreeNode, k int) int {
	var traverse func(node *TreeNode)
	var result []int
	traverse = func(node *TreeNode) {
		if node == nil {
			return
		}

		traverse(node.Left)
		result = append(result, node.Val)
		traverse(node.Right)
	}

	traverse(root)

	return result[k-1]
}

File: leetcode/230.KthSmallestElementInABST.ts
import { TreeNode } from './shared'

function kthSmallest(root: TreeNode | null, k: number): number {
  const values: number[] = []
  const dfs = (node: TreeNode | null) => {
    if (!node) return

    if (node) {
      dfs(node.left)
      values.push(node.val)
      dfs(node.right)
    }
  }

  dfs(root)

  return values[k - 1]
}
const tree = new TreeNode(3)
tree.left = new TreeNode(1)
tree.left.right = new TreeNode(2)
tree.right = new TreeNode(4)

console.log(kthSmallest(tree, 1))
